src/libsyntax/feature_gate.rs

```diff
@@ -1028,22 +987,22 @@ impl<'a, 'v> Visitor<'v> for PostExpansionVisitor<'a> {
     fn visit_expr(&mut self, e: &ast::Expr) {
         match e.node {
             ast::ExprKind::Box(_) => {
-                self.gate_feature("box_syntax",
+                gate_feature_post!(&self, box_syntax,
                                   e.span,
                                   "box expression syntax is experimental; \
                                    you can call `Box::new` instead.");
```

Comment:
nit: this can be just `gate_feature_post!(&self, box_syntax, e.span, EXPLAIN_BOX_SYNTAX);`